### PR TITLE
SU97 check for icinga and if changes is available

### PR DESCRIPTION
**Pull request by mano3567**
relates to https://github.com/sni/lmd/issues/2

### CODE REVIEW

**Comment by sni** [comment]
Great. Could you fix the format errors by running "make fmt". Btw, you can run all citests with "make citests". Can you rename the flag `Icinga2Only`, because it doesn't affect icinga 1.
I will add a few more remarks inline.

File: lmd/objects.go [hunk 1]
```diff
@@ -88,6 +88,9 @@ const (
 
 	// ShinkenOnly flag is set if the remote site is a shinken installation.
 	ShinkenOnly
+
+	// IcingaOnly flag is set if the remote site is a shinken installation.
+	IcingaOnly
 )
```
Comment (sni) [inline lmd/objects.go]: Make this `Icinga2Only` please.

File: lmd/peer.go [hunk 2]
```diff
@@ -193,6 +193,31 @@ func (p *Peer) Stop() {
...
+	if e!=nil {
+		log.Debugf("problems getting count for %s (err: %s, res: %s): %s\n", name, err, res, e)
+		count = -1
+	} 
+	return
+}
+
+func (p *Peer) hasChanged() (changed bool) {
+	changed = false
+	tablenames := []string {"commands", "comments", "contactgroups", "contacts", "downtimes", "hostgroups", "hosts", "servicegroups", "services", "timeperiods"}
```
Comment (sni) [inline lmd/peer.go]: comments and downtimes should be removed here. They can handle changed counts already.

File: lmd/peer.go [hunk 3]
```diff
@@ -193,6 +193,31 @@ func (p *Peer) Stop() {
...
+	for _, name := range tablenames {
+		counter := p.countFromServer(name, "name >= .*")
+		if counter<0  {
+			counter = p.countFromServer(name, "host_name >= .*")
```
Comment (sni) [inline lmd/peer.go]: this is quite ugly and will probably result in error messages in the livestatus.log for each check.

File: lmd/peer.go [hunk 4]
```diff
@@ -193,6 +193,31 @@ func (p *Peer) Stop() {
 	close(p.shutdownChannel)
 }
 
+func (p *Peer) countFromServer(name string, queryCondition string) (count int) {
+	err, res := p.QueryString("GET "+name+"\nStats: "+queryCondition+"\n\n") 
+	count, e := strconv.Atoi(fmt.Sprintf("%s",res))
+	if e!=nil {
+		log.Debugf("problems getting count for %s (err: %s, res: %s): %s\n", name, err, res, e)
```
Comment (sni) [inline lmd/peer.go]: please use the standard logformat like: `log.Debugf("[%s] ...", p.Name)`

**Comment by mano3567** [comment]
thanks for the feedback, i think i addressed them a moment ago.
best regards magnus

File: lmd/peer.go [hunk 5]
```diff
@@ -1097,15 +1125,20 @@ func (p *Peer) createIndexAndFlags(table *Table, res *[][]interface{}, index *ma
...
 		matched := reShinkenVersion.FindStringSubmatch(row[table.GetColumn("livestatus_version").Index].(string))
 		if len(matched) > 0 {
 			p.PeerLock.Lock()
 			p.Flags |= ShinkenOnly
 			p.PeerLock.Unlock()
-		}
+		} 
+		if len(reIcingaVersion.FindStringSubmatch(row[table.GetColumn("livestatus_version").Index].(string))) > 0 {
+			p.PeerLock.Lock()
+			p.Flags |= IcingaOnly
```
Comment (sni) [inline lmd/peer.go]: i think you forgot one to rename. please run `make citest` :-)

File: lmd/peer.go [hunk 6]
```diff
@@ -228,6 +250,12 @@ func (p *Peer) updateLoop() {
 					p.UpdateObjectByType(Objects.Tables["hostgroups"])
 					p.UpdateObjectByType(Objects.Tables["servicegroups"])
 					lastTimeperiodUpdateMinute = currentMinute
+					if p.Flags&Icinga2Only == Icinga2Only {
+						if p.hasChanged() {
+							ok = p.InitAllTables()
+							lastTimeperiodUpdateMinute = currentMinute
```
Comment (sni) [inline lmd/peer.go]: i'd say this line can be removed, it is set a few lines earlier already.

File: lmd/peer.go [hunk 7]
```diff
@@ -193,6 +193,28 @@ func (p *Peer) Stop() {
...
+	return
+}
+
+func (p *Peer) hasChanged() (changed bool) {
+	changed = false
+	tablenames := []string {"commands", "contactgroups", "contacts", "hostgroups", "hosts", "servicegroups", "services", "timeperiods"}
+	for _, name := range tablenames {
+		counter := p.countFromServer(name, "name >= .*")
+		if counter<0  {
+			counter = p.countFromServer(name, "host_name >= .*")
```
Comment (sni) [inline lmd/peer.go]: this is still ugly :-)

**Comment by sni** [comment]
Thanks, i just added some more notes.

**Comment by mano3567** [comment]
so - i just made some changes, hope it looks better :)

**Comment by sni** [comment]
Looks good. I will merge it tomorrow. I'd still change the filter. >= .* may work on Icinga 2 but is wrong, `>=` is used for lists or number, not on names. So i'd suggest using `!= ` to match everything. The failed unit test is somethin if have to look into...

**Comment by sni** [comment]
Thanks, i've merged it manually and did some minor changes.

**Comment by hkowallik** [comment]
I seems icinga2 changed the version string for livestatus.
On a current 2.6.3 version, I get "v2.6.3" which doesn't match the regular expression used to detect icinga2.
var reIcingaVersion = regexp.MustCompile(`^r[\d\.-]+$`)

for me changing this to
var reIcingaVersion = regexp.MustCompile(`^v2\.`)
helps but I don't know that other version strings are out there so I can't suggest a regex that will always work

**Comment by sni** [comment]
maybe you could ask them to append a "-icinga2" at the datasource version in their livestatus?

**Pull request closed without merging**